processing/src/main/java/org/apache/druid/query/monomorphicprocessing/SpecializationService.java

```diff
@@ -65,19 +69,162 @@
     try {
-      Field theUnsafe = Unsafe.class.getDeclaredField("theUnsafe");
-      theUnsafe.setAccessible(true);
-      UNSAFE = (Unsafe) theUnsafe.get(null);
+      defineClass = lookupDefineClassMethodHandle();
     }
-    catch (Exception e) {
-      throw new RuntimeException("Cannot access Unsafe methods", e);
+    catch (RuntimeException e) {
+      exception = e;
+    }
+    if (defineClass != null) {
+      DEFINE_CLASS = defineClass;
+      DEFINE_CLASS_NOT_SUPPORTED_EXCEPTION = null;
+    } else {
+      DEFINE_CLASS = null;
+      DEFINE_CLASS_NOT_SUPPORTED_EXCEPTION = exception;
     }
   }
 
+  private static MethodHandle lookupDefineClassMethodHandle()
+  {
+    try {
+      MethodHandles.Lookup lookup = MethodHandles.lookup();
+      if (JvmUtils.isIsJava9Compatible()) {
+        return defineClassJava9(lookup);
+      } else {
+        return defineClassJava8(lookup);
+      }
+    }
+    catch (ReflectiveOperationException | RuntimeException e) {
+      throw new UnsupportedOperationException(
+          "defineClass is not supported on this platform, "
+          + "because internal Java APIs are not compatible with this Druid version",
+          e
+      );
+    }
+  }
+
+  private static MethodHandle defineClassJava9(MethodHandles.Lookup lookup) throws ReflectiveOperationException
+  {
+    // "Compile" a MethodHandle that is equivalent to the following closure:
+    //
+    //  Class<?> defineClass(Class targetClass, String className, byte[] byteCode) {
+    //    MethodHandles.Lookup targetClassLookup = MethodHandles.privateLookupIn(targetClass, lookup);
+    //    return targetClassLookup.defineClass(byteCode);
+    //  }
+
+    // this is getting meta
+    MethodHandle defineClass = lookup.unreflect(MethodHandles.Lookup.class.getMethod("defineClass", byte[].class));
+    MethodHandle privateLookupIn = lookup.findStatic(
+        MethodHandles.class,
+        "privateLookupIn",
+        MethodType.methodType(MethodHandles.Lookup.class, Class.class, MethodHandles.Lookup.class)
+    );
+
+    // bind privateLookupIn lookup argument to this method's lookup
+    // privateLookupIn = (Class targetClass) -> privateLookupIn(MethodHandles.privateLookupIn(targetClass, lookup))
+    privateLookupIn = MethodHandles.insertArguments(privateLookupIn, 1, lookup);
+
+    // defineClass = (Class targetClass, byte[] byteCode) -> privateLookupIn(targetClass).defineClass(byteCode)
+    defineClass = MethodHandles.filterArguments(defineClass, 0, privateLookupIn);
+
+    // add a dummy String argument to match the corresponding JDK8 version
+    // defineClass = (Class targetClass, byte[] byteCode, String className) -> defineClass(targetClass, byteCode)
+    defineClass = MethodHandles.dropArguments(defineClass, 2, String.class);
+    return defineClass;
+  }
+
+  private static MethodHandle defineClassJava8(MethodHandles.Lookup lookup) throws ReflectiveOperationException
+  {
+    Class<?> unsafeClass = Class.forName("sun.misc.Unsafe");
```

Comment:
I think it's still useful to have a central registry that holds a reference to `Unsafe` (the registry class is loaded only when running on Java 8)